### ISSUE TITLE
Custom BaseTransform not reflecting changes in train_batch images

### ISSUE THREAD

**Issue by HeavensDoor1** [ultralytics/ultralytics#23942]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi Ultralytics team,

I’m implementing a custom augmentation for a segmentation task, following the logic in [Issue #20596](https://github.com/ultralytics/ultralytics/issues/20596).  I'm trying to downscale specific objects (Class 1) and place them on a fixed-size canvas (imgsz) within the augmentation pipeline.

### Use case

Despite updating labels["img"] and labels["instances"], the train_batch*.jpg files still show original large objects. It seems Mosaic or the caching system bypasses these early changes.

```
import cv2
import numpy as np
from ultralytics.data.augment import BaseTransform, v8_transforms
from ultralytics import YOLO

class LimitObjectSize(BaseTransform):
    def __init__(self, max_obj_size=30, target_cls=1, imgsz=480):
        super().__init__()
        self.max_obj_size, self.target_cls, self.imgsz = max_obj_size, target_cls, imgsz

    def __call__(self, labels):
        instances = labels.get("instances")
        if instances is None or len(instances) == 0: return labels
        
        mask = (instances.cls == self.target_cls)
        if not mask.any(): return labels

        # Calculate scale to fit max_obj_size
        bboxes = instances.bboxes[mask]
        current_max = max((bboxes[:, 2] - bboxes[:, 0]).max(), (bboxes[:, 3] - bboxes[:, 1]).max())
        
        if current_max > self.max_obj_size:
            scale = self.max_obj_size / current_max
            img = cv2.resize(labels["img"], (int(labels["img"].shape[1] * scale), int(labels["img"].shape[0] * scale)))
            
            # Create imgsz canvas and center the resized image
            final_img = np.zeros((self.imgsz, self.imgsz, 3), dtype=np.uint8)
            pad_x, pad_y = (self.imgsz - img.shape[1]) // 2, (self.imgsz - img.shape[0]) // 2
            final_img[pad_y:pad_y+img.shape[0], pad_x:pad_x+img.shape[1]] = img

            # Update geometry
            instances.scale(scale, scale)
            instances.convert_bbox('xyxy')
            instances.bboxes += np.array([pad_x, pad_y, pad_x, pad_y], dtype=np.float32)
            if hasattr(instances, 'segments'):
                for i in range(len(instances.segments)): instances.segments[i] += np.array([pad_x, pad_y], dtype=np.float32)

            labels.update({"img": final_img, "resized_shape": (self.imgsz, self.imgsz)})
        return labels
def custom_transforms_adapter(dataset, imgsz, hyp):
    T = v8_transforms(dataset, imgsz, hyp)
    T.transforms.insert(0, LimitObjectSize(imgsz=imgsz))
    return T
# Callback registration
model = YOLO('yolo11n-seg.pt')
model.add_callback('on_train_start', lambda trainer: setattr(trainer.data, 'transforms', custom_transforms_adapter))
model.train(data='data.yaml', imgsz=480)

```

How can I ensure my geometric modifications are respected by the Mosaic augmentation?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @HeavensDoor1, thank you for reporting this and for sharing the detailed context around your custom `BaseTransform` for segmentation 🚀 This is an automated response to help get the issue triaged quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible. In particular, a minimal script and tiny sample dataset that reproduces the `train_batch*.jpg` behavior would help us investigate whether this is related to transforms order, Mosaic, or caching behavior 🔍

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
hi @HeavensDoor1, three issues here:

1. **wrong target** — `trainer.data` is a `dict`, not the dataset. transforms live at `trainer.train_loader.dataset.transforms`

2. **wrong position** — inserting at position 0 puts your transform *before* mosaic, but mosaic then creates a new composite image from 4 raw images, overwriting your changes. insert at position **1** (after mosaic) instead

3. **instances API** — after mosaic, classes are in `labels["cls"]` not `instances.cls`, and `instances.bboxes` is read-only — use `instances.scale()` / `instances.add_padding()` for geometry updates

**Comment by HeavensDoor1** [comment]
> hi [@HeavensDoor1](https://github.com/HeavensDoor1), three issues here:
> 
> 1. **wrong target** — `trainer.data` is a `dict`, not the dataset. transforms live at `trainer.train_loader.dataset.transforms`
> 2. **wrong position** — inserting at position 0 puts your transform _before_ mosaic, but mosaic then creates a new composite image from 4 raw images, overwriting your changes. insert at position **1** (after mosaic) instead
> 3. **instances API** — after mosaic, classes are in `labels["cls"]` not `instances.cls`, and `instances.bboxes` is read-only — use `instances.scale()` / `instances.add_padding()` for geometry updates


I reworked the code according to your comments, but still nothing changes in the train batches.

```
import cv2
import numpy as np
from ultralytics.data.augment import BaseTransform, v8_transforms

class LimitObjectSize(BaseTransform):
    def __init__(self, max_obj_size=30, target_cls=1, imgsz=480):
        super().__init__()
        self.max_obj_size = max_obj_size
        self.target_cls = target_cls
        self.imgsz = imgsz

    def __call__(self, labels):
        instances = labels.get("instances")
        cls = labels.get("cls")
        
        if instances is None or cls is None or len(cls) == 0:
            return labels

        mask = (cls.flatten() == self.target_cls)
        if not mask.any():
            return labels

        target_bboxes = instances.bboxes[mask]
        if len(target_bboxes) == 0:
            return labels

        widths = target_bboxes[:, 2] - target_bboxes[:, 0]
        heights = target_bboxes[:, 3] - target_bboxes[:, 1]
        current_max = max(widths.max(), heights.max())

        if current_max <= self.max_obj_size:
            return labels

        scale = self.max_obj_size / current_max
        img = labels["img"]
        h, w = img.shape[:2]
        new_w, new_h = int(w * scale), int(h * scale)
        
        img_resized = cv2.resize(img, (new_w, new_h), interpolation=cv2.INTER_AREA)

        pad_left = (self.imgsz - new_w) // 2
        pad_top = (self.imgsz - new_h) // 2
        pad_right = self.imgsz - new_w - pad_left
        pad_bottom = self.imgsz - new_h - pad_top

        canvas = np.zeros((self.imgsz, self.imgsz, 3), dtype=np.uint8)
        canvas[pad_top:pad_top+new_h, pad_left:pad_left+new_w] = img_resized

        instances.scale(scale, scale)
        instances.add_padding(pad_left, pad_top, pad_right, pad_bottom)

        labels["img"] = canvas
        labels["resized_shape"] = (self.imgsz, self.imgsz)
        
        return labels

def setup_custom_pipeline(trainer):
    dataset = trainer.train_loader.dataset
    imgsz = trainer.args.imgsz
    hyp = trainer.args
    
    T = v8_transforms(dataset, imgsz, hyp)
    T.transforms.insert(1, LimitObjectSize(max_obj_size=20, target_cls=1, imgsz=imgsz))
    
    dataset.transforms = T
```

**Comment by raimbekovm** [comment]
@HeavensDoor1 the issue is the insertion point. `v8_transforms` returns this structure:

```
Compose([                          # outer
    Compose([                      # pre_transform
        Mosaic,
        CopyPaste,
        RandomPerspective(LetterBox),  # LetterBox resizes to imgsz here
    ]),
    MixUp, CutMix, Albumentations, RandomHSV, RandomFlip, RandomFlip
])
```

`transforms.transforms.insert(1, ...)` inserts into the **outer** Compose — after `pre_transform`, where the image is already `480×480` (LetterBox already ran). your scaling condition never triggers because objects are already small at that point.

the fix is to insert into `pre_transform[1]` — after Mosaic but before RandomPerspective/LetterBox. at that point the image is a raw mosaic at `960×960` and objects are at their original pixel scale. i verified this — the custom transform correctly sees `960×960` images when inserted there.

working approach:

```python
import cv2
import numpy as np
from ultralytics import YOLO
from ultralytics.data.augment import BaseTransform, Compose
from ultralytics.data.dataset import YOLODataset
from ultralytics.models.yolo.detect.train import DetectionTrainer


class LimitObjectSize(BaseTransform):
    def __init__(self, max_obj_size=30, target_cls=1, imgsz=480):
        super().__init__()
        self.max_obj_size = max_obj_size
        self.target_cls = target_cls
        self.imgsz = imgsz

    def __call__(self, labels):
        instances = labels.get("instances")
        cls = labels.get("cls")
        if instances is None or len(instances) == 0 or cls is None:
            return labels

        mask = cls.flatten() == self.target_cls
        if not mask.any():
            return labels

        target_bboxes = instances.bboxes[mask]
        widths = target_bboxes[:, 2] - target_bboxes[:, 0]
        heights = target_bboxes[:, 3] - target_bboxes[:, 1]
        current_max = max(widths.max(), heights.max())

        if current_max <= self.max_obj_size:
            return labels

        scale = self.max_obj_size / current_max
        h, w = labels["img"].shape[:2]
        new_w, new_h = int(w * scale), int(h * scale)
        labels["img"] = cv2.resize(labels["img"], (new_w, new_h))

        canvas = np.zeros((self.imgsz, self.imgsz, 3), dtype=np.uint8)
        pad_x = (self.imgsz - new_w) // 2
        pad_y = (self.imgsz - new_h) // 2
        end_x = min(pad_x + new_w, self.imgsz)
        end_y = min(pad_y + new_h, self.imgsz)
        canvas[pad_y:end_y, pad_x:end_x] = labels["img"][: end_y - pad_y, : end_x - pad_x]

        instances.scale(scale, scale)
        instances.add_padding(pad_x, pad_y, self.imgsz - new_w - pad_x, self.imgsz - new_h - pad_y)
        labels["img"] = canvas
        return labels


class CustomDataset(YOLODataset):
    def build_transforms(self, hyp=None):
        transforms = super().build_transforms(hyp)
        if self.augment and isinstance(transforms, Compose) and len(transforms.transforms) > 0:
            pre = transforms.transforms[0]
            if isinstance(pre, Compose):
                pre.insert(1, LimitObjectSize(imgsz=self.imgsz))
        return transforms


_original = DetectionTrainer.build_dataset


def _custom_build(self, img_path, mode="train", batch=None):
    gs = max(int(self.model.stride.max() if hasattr(self.model, "stride") else 32), 32)
    return CustomDataset(
        img_path=img_path,
        imgsz=self.args.imgsz,
        batch_size=batch,
        augment=mode == "train",
        hyp=self.args,
        rect=self.args.rect if mode == "train" else True,
        cache=self.args.cache,
        single_cls=self.args.single_cls,
        stride=gs,
        pad=0.0 if mode == "train" else 0.5,
        prefix=f"{mode}: ",
        task=self.args.task,
        classes=self.args.classes,
        data=self.data,
        fraction=self.args.fraction if mode == "train" else 1.0,
    )


DetectionTrainer.build_dataset = _custom_build
model = YOLO("yolo11n-seg.pt")
model.train(data="data.yaml", imgsz=480)
DetectionTrainer.build_dataset = _original
```

note: since your transform runs before LetterBox, the image will still be resized to `imgsz` afterwards. make sure your scaling logic accounts for the raw mosaic dimensions (typically `2 × imgsz`).

**Comment by HeavensDoor1** [comment]
Thanks for the help, everything is working now! I have a follow-up question: Currently, we are resizing the entire Mosaic canvas when a target object is too large. Is there a way within the Ultralytics framework to resize only the specific image (tile) within the Mosaic where the object exceeds the size limit, rather than scaling down the whole 4nd-image composition?

**Comment by glenn-jocher** [comment]
Yes, but not as a post-Mosaic `BaseTransform`—once [`Mosaic`](https://docs.ultralytics.com/reference/data/augment/#ultralyticsdataaugmentmosaic) finishes, `_mosaic4()` has already placed the 4 tiles and `_cat_labels()` has merged their labels, so tile identity is effectively gone; if you want to resize only one source tile, the clean approach is to subclass `Mosaic` (or `BaseMixTransform`) and modify the selected `labels`/`mix_labels[i]` before that tile is composited.
